### ISSUE TITLE
Speed up PPO training loop

### ISSUE THREAD

**Issue by cswinter** [entity-neural-network/incubator#12]
There are still many performance issues with the PPO and agent implementation, on my machine I'm only getting 8 steps per second. Since actions/logprobs/observations/etc. don't have fixed shapes, they are all collected in Python lists and not currently vectorized. We should profile and (very carefully, without adding any bugs) eliminate inefficiencies.

**Comment by cswinter** [comment]
Currently, the majority of the time seems to be taken up by this line:

https://github.com/entity-neural-network/incubator/blob/e7057884a005557d1d0ef5bfd9467cca59cedd03/ppo/train.py#L167

Seems there is a very high overhead to creating a PyTorch `Categorical` object. I'm also seeing W&B take up another good chunk of time, unclear why that should be the case.

Flamegraph of profile recorded with py-spy: https://www.dropbox.com/s/7d04oeh4y9x0wdj/2021-11-23%20ppo-flamegraph.svg?dl=0

**Comment by cswinter** [comment]
https://github.com/entity-neural-network/incubator/pull/15 gives speedup from 8 SPS to 260 SPS, mainly by batching forward passes and only creating one `Categorical` object per action and forward pass 😎
Still a lot further to go but this is workable now.
The implementation currently also only works with categorical actions, the `SelectEntity` action will probably require a custom ragged softmax kernel to be efficient.

**Comment by cswinter** [comment]
Up to 10x speedup with https://github.com/entity-neural-network/incubator/pull/40, with enough parallel envs > 10K is possible.

**Comment by cswinter** [comment]
Further speedup with https://github.com/entity-neural-network/incubator/pull/81pushing us to > 14K/s on `MultiArmedBandit` with 64 envs. While many more optimizations are possible, all of the inefficiencies caused by ragged observation/action spaces are now addressed, and the environment and network architecture already ought to be the bottleneck in most cases.
